megatron/data/ul2_dataset.py

```diff
@@ -0,0 +1,283 @@
+
+    assert target_seq_length <= max_seq_length
+
+    # flatten sentences into one list
+    tokens = [token for sentence in sample for token in sentence]
+
+    max_num_tokens = target_seq_length
+    # if is_decoder_only(model_type):
+    #     # Keep space for repeated `extra_id` tokens; not the most data
+    #     # efficient since we calculate this based on the maximum number
+    #     # of possible `extra_id` tokens.
+    #     safe_max_seq_len = math.floor(max_num_tokens / (1 + masked_lm_prob))
+    #     truncated = len(tokens) > safe_max_seq_len
+    #     tokens = tokens[:safe_max_seq_len]
+    # else:
+    # Truncate to `target_sequence_length`.
+    truncated = len(tokens) > max_num_tokens
+    tokens = tokens[:max_num_tokens]
+
+    # Prepend objective token.
+    cls_id = cls_ids.get(denoiser)
+    if cls_id is None:
+        raise ValueError('unknown denoiser')
+    tokens = [cls_id] + tokens
+
+    # Masking.
+    mean_ngrams = mean_span_lengths[denoiser_index]
+    if mean_ngrams < 1:
+        # Ensure we always obtain at least one `max_ngrams`.
+        mean_ngrams = max(1, round(len(tokens) * mean_ngrams))
+    max_ngrams = mean_ngrams * 2 - 1
+
+    if denoiser == 'R' or denoiser == 'X':
+        sampling_style = SamplingStyle.NORMAL
+        prefix_lm = False
+        max_predictions_per_seq = len(tokens) - 1
+    elif denoiser == 'S':
+        sampling_style = SamplingStyle.UNIFORM
+        prefix_lm = True
+        max_predictions_per_seq = min(
+            round(masked_lm_prob * len(tokens)) * 2 - 1,
+            len(tokens) - 1,
+        )
+    else:
+        raise ValueError('unknown denoiser')
+
+    # Ensure we always have at least one prediction.
+    max_predictions_per_seq = max(1, max_predictions_per_seq)
+    (
+        tokens, masked_positions, masked_labels, _, masked_spans,
+    ) = create_masked_lm_predictions(
+        tokens, vocab_id_list, vocab_id_to_token_dict, masked_lm_prob,
+        cls_id, sep_id, mask_id, max_predictions_per_seq, np_rng,
+        max_ngrams=max_ngrams, masking_style="t5",
+        sampling_style=sampling_style, prefix_lm=prefix_lm,
+    )
+
+    if is_decoder_only(model_type):
+        # Concatenate to one sequence.
+        tokens_enc, tokens_dec_in, labels = merge_subsequent_masks(
+            tokens, masked_spans, bos_id, eos_id, sentinel_tokens, prefix_lm)
+
+        # Move EOS tokens to end of sequence.
+        while tokens_enc[-1] == eos_id:
+            del tokens_enc[-1]
+            tokens_dec_in.append(eos_id)
+            labels.append(eos_id)
+
+        num_labels = len(labels)
+
+        # Move BOS token to start of sequence.
+        tokens_dec_in = tokens_dec_in[1:]
+        tokens = (
+            [bos_id]
+            + tokens_enc
+            + [sep_id]
+            + tokens_dec_in
+        )
+
+        # Pad and convert to NumPy.
+        padding_length = max_seq_length - len(tokens)
+        if padding_length < 0:
+            raise LengthExceededError()
```

Comment:
Maybe I'm missing something, but I have some trouble understanding how this exception does not occur almost always.
Each masked span adds an extra token in `tokens_enc` and in `tokens_dec_in`.
So the concatenation `tokens = (
            [bos_id]
            + tokens_enc
            + [sep_id]
            + tokens_dec_in
        )` should be longer than `max_seq_length` ?

Reply:
In my case decoder-only I uncommented 
```
    # if is_decoder_only(model_type):
    #     # Keep space for repeated `extra_id` tokens; not the most data
    #     # efficient since we calculate this based on the maximum number
    #     # of possible `extra_id` tokens.
    #     safe_max_seq_len = math.floor(max_num_tokens / (1 + masked_lm_prob))
    #     truncated = len(tokens) > safe_max_seq_len
    #     tokens = tokens[:safe_max_seq_len]
    # else:
```
& then it worked; Just for ref the code I'm using is https://github.com/TurkuNLP/Megatron-DeepSpeed/blob/main/megatron/data/ul2_dataset.py

Reply:
Each masked span adds an extra token in `tokens_dec_in`. `tokens_enc` is actually reduced in size since multiple tokens are replaced by a single mask.
However, you're right, `len(tokens_enc + tokens_dec_in)` increases by 1 per span mask.

For the decoder-only case, this is indeed too strict and should be handled better than asking the user to specify a sequence length larger than any text in the data set.

While I still dislike my hacky solution that Niklas cited for the reasons I wrote in the comment, it's probably best to keep it (with a few adjustments due to the recent changes, which can still cause the seq length to overflow).

Ideally, there would be a solution that fixes the length excess problems even for the original `T5Dataset`, since it can still happen in the encoder-decoder case as well.